numpy/core/_exceptions.py

```diff
@@ -122,21 +122,89 @@ class TooHardError(RuntimeError):
 
 @set_module('numpy')
 class AxisError(ValueError, IndexError):
-    """ Axis supplied was invalid. """
+    """Axis supplied was invalid.
+
+    Raised whenever an ``axis`` parameter is specified that is larger than
+    the number of array dimensions.
+
+    .. versionadded:: 1.22
+
+        Added the ``axis`` and ``ndim`` attributes.
+
+    Parameters
+    ----------
+    axis : int or str
+        The out of bounds axis or a custom exception message.
+        If an axis is provided, then `ndim` should be specified as well.
+    ndim : int, optional
+        The number of array dimensions.
+    msg_prefix : str, optional
+        A prefix for the exception message.
+
+    Attributes
+    ----------
+    axis : int, optional
+        The out of bounds axis or ``None`` if a custom exception
+        message was provided.
```

Comment:
```suggestion
        message was provided. This should be the axis as passed by
        the user, before any normalization to resolve negative indices.
```

Reply:
Fixed as of https://github.com/numpy/numpy/pull/19459/commits/38b0fc25b4bc054d399ecfe54e30596888cda21e.